flink/src/main/java/org/apache/iceberg/flink/writer/OutputFileFactory.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.writer;
+
+import java.util.UUID;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.flink.PartitionKey;
+import org.apache.iceberg.io.OutputFile;
+
+public class OutputFileFactory {
+  private final String uuid = UUID.randomUUID().toString();
+
+  private final Table table;
+  private final FileFormat format;
+  private final long taskId;
+  private int fileCount;
+
+  public OutputFileFactory(Table table, FileFormat format, long taskId) {
```

Comment:
`Table` is not serializable. That's why the Spark output file factory has fields for `LocationProvider`, `FileIO`, `EncryptionManager`, and `PartitionSpec` that are serializable.

Reply:
OK, I think I get the difference you mentioned between spark writer implementation and flink writer implementation:
For spark,   it will load the iceberg table firstly at driver side, then create the `DataWriterFactory`  and serialize and dispatch it to each executor, then the executor will create its `DataWriter`, so here each executor won't need to load the iceberg table.   
For flink , I currently implemented the `IcebergStreamWriter` by loading the iceberg table for each sub task, then each task get the table path and open the iceberg table, so the `Table` instance won't need to be serializable. Seems the iceberg table will be loaded 100 times if we have 100 parallerism for `IcebergStreamWriter`.  @JingsongLi Any thought about this issue ?  I mean: do we have some similar ways as spark did to optimize the iceberg table loading ?

Reply:
I don't quite understand what you mean.
First, the `StreamOperator.open` is invoked in tasks instead of client.

Maybe we should avoid loading table in tasks, image the catalog is hive catalog, too many tasks visit the catalog will lead to collapse of hive metastore.
So in this direction, we should store serializable `LocationProvider`, `FileIO`, `EncryptionManager`, and `PartitionSpec` in `StreamOperator` instead of loading table at runtime.

Reply:
I'd prefer to avoid loading the table in each task. There is no need for the table reference on workers, and not having it will prevent patterns that are bad, like attempting to commit to the table from every task.

Reply:
Agree.   I changed the design in commit https://github.com/apache/iceberg/pull/1145/commits/0026d1676f1319ce9442136c8934c015ced39aa0, say it will load the table firstly, then initialize the `IcebergStreamWriter`  instances and then deploy the flink stream job to flink jobManager, and finally the flink jobManager will allocate slot and dispatch the task to each slot. In general,  we currently load the table once.. Thanks for the discussion.